array.go

```diff
@@ -1193,3 +1193,143 @@ func countElement(array []interface{}, element interface{}) int {
 	}
 	return count
 }
+
+// IsOrdered succeeds if array is ordered based on optional `comparator` function.
+// For default, it will use built-in comparator function for each data type.
+// It requires all elements in the array to have same data type
+//
+// Example:
+//
+//	array := NewArray(t, []interface{}{100, 101, 102})
+//	array.IsOrdered(func(x, y *Value) bool {
+//		valX, _ := x.Raw().(int)
+//		valY, _ := y.Raw().(int)
+//		return valX >= valY
+//	})
+func (a *Array) IsOrdered(less ...func(x, y *Value) bool) *Array {
+	a.chain.enter("IsOrdered()")
+	defer a.chain.leave()
+
+	if a.chain.failed() {
+		return a
+	}
+
+	if len(less) > 1 {
+		a.chain.fail(AssertionFailure{
+			Type: AssertUsage,
+			Errors: []error{
+				errors.New("unexpected multiple less arguments"),
+			},
+		})
+		return a
+	}
+
+	if len(a.value) == 0 {
+		return a
+	}
+
+	var fn func(x, y *Value) bool
+	if len(less) == 1 {
+		fn = less[0]
+	} else {
+		// use default comparator based on data type
+		cArr, ok := canonArray(a.chain, a.value)
+		if !ok {
+			return a
+		}
+
+		fn = a.getDefaultComparator(cArr)
+		if a.chain.failed() {
+			return a
+		}
+	}
+
+	for i := 0; i < int(a.Length().Raw())-1; i++ {
+		xChain := a.chain.clone()
+		xChain.replace("IsOrdered[%d]", i)
+		x := newValue(xChain, a.value[i])
+
+		yChain := a.chain.clone()
+		yChain.replace("IsOrdered[%d]", i+1)
+		y := newValue(yChain, a.value[i+1])
+
+		if !fn(x, y) {
+			a.chain.fail(AssertionFailure{
+				Type:      AssertLt,
+				Actual:    &AssertionValue{x.value},
+				Expected:  &AssertionValue{y.value},
+				Reference: &AssertionValue{a.value},
+				Errors: []error{
+					errors.New("expected: array is ordered"),
+					fmt.Errorf("element %v must be less than element %v, but it isn't",
+						x.value, y.value),
+				},
+			})
+			return a
+		}
+	}
+
+	return a
+}
+
+func (a *Array) getDefaultComparator(cArr []interface{}) func(x, y *Value) bool {
+	var fn func(x, y *Value) bool
+	var currType, prevType string
+	var prevElem interface{}
+	for _, v := range cArr {
+		switch t := v.(type) {
+		case bool:
+			currType = fmt.Sprintf("%T", t)
+			fn = func(x, y *Value) bool {
```

Comment:
This code will recreate fn lambda each iteration, which is sub-optimal. Instead, I suggest only to check types in the loop, and then create fn outside of the loop.

Reply:
> done, also I changed the builtin comparator to use httpexpect functionality, instead of golang assertion

In builtin comparator, it's better to avoid using nested assertions, as mentioned here: https://github.com/gavv/httpexpect/pull/214#discussion_r1058522115

Because:

* user do not expect those assertions (there are no of them in user code)
* user will have misleading error message

On the other hand, in the example it's OK to use them. If the user explicitly provides custom func that does those assertions, they are not misleading and are completely expected.

Reply:
getBuiltinComparator looks quite lengthy; maybe split it into two functions? one for type check, another for generating comparator function.